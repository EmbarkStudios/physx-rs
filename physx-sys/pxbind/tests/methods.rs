struct Functions {
    cpp: String,
    rust: String,
}

fn gen_functions(which: &str, to_emit: &'static [&str]) -> anyhow::Result<Functions> {
    let ast = pxbind::get_parsed_ast(format!("tests/data/functions/{which}"))?;

    let mut consumer = pxbind::consumer::AstConsumer::default();
    consumer.consume(&ast)?;

    let record_filter = |_rb: &pxbind::consumer::RecBinding<'_>| false;
    let enum_filter = |_eb: &pxbind::consumer::EnumBinding<'_>| false;
    let func_filter = |fb: &pxbind::consumer::FuncBinding<'_>| {
        if let Some(cname) = fb.owning_class() {
            to_emit.is_empty() || to_emit.iter().any(|te| *te == cname)
        } else {
            false
        }
    };

    let generator = pxbind::generator::Generator {
        record_filter: Box::new(record_filter),
        enum_filter: Box::new(enum_filter),
        func_filter: Box::new(func_filter),
        ..Default::default()
    };

    let cpp = {
        let mut buffer = Vec::new();
        generator.generate_cpp_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    let rust = {
        let mut buffer = Vec::new();
        generator.generate_rust_functions(&consumer, &mut buffer, 0)?;
        String::from_utf8(buffer)?
    };

    Ok(Functions { cpp, rust })
}

/// Verifies that reference fields are reflected in the size and layout, but aren't
/// actually publicly exposed
#[test]
fn methods_complex() {
    let mo = gen_functions("methods_complex.h", &["PxShape", "PxThreadImpl"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies constructors of various kinds are properly bound, and properly
/// named to avoid clashes
#[test]
fn constructors() {
    let mo = gen_functions("constructors.h", &["PxBoxControllerDesc", "PxFilterData"]).unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}

/// Verifies destructors are properly emitted or omitted
#[test]
fn destructor() {
    let mo = gen_functions(
        "destructors.h",
        &[
            "PxActor",
            "PxJointLimitParameters",
            "PxJointAngularLimitPair",
        ],
    )
    .unwrap();

    insta::assert_snapshot!(mo.cpp);
    insta::assert_snapshot!(mo.rust);
}
