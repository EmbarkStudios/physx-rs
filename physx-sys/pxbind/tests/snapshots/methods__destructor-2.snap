---
source: physx-sys/pxbind/tests/methods.rs
expression: mo.rust
---
extern "C" {
    /// Deletes the actor.
    ///
    /// Do not keep a reference to the deleted instance.
    ///
    /// If the actor belongs to a [`PxAggregate`] object, it is automatically removed from the aggregate.
    pub fn PxActor_release_mut(self_: *mut physx_PxActor_Pod);

    /// Retrieves the type of actor.
    ///
    /// The actor type of the actor.
    pub fn PxActor_getType(self_: *const physx_PxActor_Pod) -> PxActorType;

    /// Retrieves the scene which this actor belongs to.
    ///
    /// Owner Scene. NULL if not part of a scene.
    pub fn PxActor_getScene(self_: *const physx_PxActor_Pod) -> *mut physx_PxScene_Pod;

    /// Sets a name string for the object that can be retrieved with getName().
    ///
    /// This is for debugging and is not used by the SDK. The string is not copied by the SDK,
    /// only the pointer is stored.
    ///
    /// Default:
    /// NULL
    pub fn PxActor_setName_mut(self_: *mut physx_PxActor_Pod, name: *const std::ffi::c_char);

    /// Retrieves the name string set with setName().
    ///
    /// Name string associated with object.
    pub fn PxActor_getName(self_: *const physx_PxActor_Pod) -> *const std::ffi::c_char;

    /// Retrieves the axis aligned bounding box enclosing the actor.
    ///
    /// It is not allowed to use this method while the simulation is running (except during PxScene::collide(),
    /// in PxContactModifyCallback or in contact report callbacks).
    ///
    /// The actor's bounding box.
    pub fn PxActor_getWorldBounds(self_: *const physx_PxActor_Pod, inflation: f32) -> physx_PxBounds3_Pod;

    /// Raises or clears a particular actor flag.
    ///
    /// See the list of flags [`PxActorFlag`]
    ///
    /// Sleeping:
    /// Does
    /// NOT
    /// wake the actor up automatically.
    pub fn PxActor_setActorFlag_mut(self_: *mut physx_PxActor_Pod, flag: PxActorFlag, value: bool);

    /// Sets the actor flags.
    ///
    /// See the list of flags [`PxActorFlag`]
    pub fn PxActor_setActorFlags_mut(self_: *mut physx_PxActor_Pod, inFlags: PxActorFlags);

    /// Reads the PxActor flags.
    ///
    /// See the list of flags [`PxActorFlag`]
    ///
    /// The values of the PxActor flags.
    pub fn PxActor_getActorFlags(self_: *const physx_PxActor_Pod) -> PxActorFlags;

    /// Assigns dynamic actors a dominance group identifier.
    ///
    /// PxDominanceGroup is a 5 bit group identifier (legal range from 0 to 31).
    ///
    /// The PxScene::setDominanceGroupPair() lets you set certain behaviors for pairs of dominance groups.
    /// By default every dynamic actor is created in group 0.
    ///
    /// Default:
    /// 0
    ///
    /// Sleeping:
    /// Changing the dominance group does
    /// NOT
    /// wake the actor up automatically.
    pub fn PxActor_setDominanceGroup_mut(self_: *mut physx_PxActor_Pod, dominanceGroup: u8);

    /// Retrieves the value set with setDominanceGroup().
    ///
    /// The dominance group of this actor.
    pub fn PxActor_getDominanceGroup(self_: *const physx_PxActor_Pod) -> u8;

    /// Sets the owner client of an actor.
    ///
    /// This cannot be done once the actor has been placed into a scene.
    ///
    /// Default:
    /// PX_DEFAULT_CLIENT
    pub fn PxActor_setOwnerClient_mut(self_: *mut physx_PxActor_Pod, inClient: u8);

    /// Returns the owner client that was specified at creation time.
    ///
    /// This value cannot be changed once the object is placed into the scene.
    pub fn PxActor_getOwnerClient(self_: *const physx_PxActor_Pod) -> u8;

    /// Retrieves the aggregate the actor might be a part of.
    ///
    /// The aggregate the actor is a part of, or NULL if the actor does not belong to an aggregate.
    pub fn PxActor_getAggregate(self_: *const physx_PxActor_Pod) -> *mut physx_PxAggregate_Pod;

    pub fn PxJointLimitParameters_new_alloc() -> *mut physx_PxJointLimitParameters_Pod;

    /// Returns true if the current settings are valid.
    ///
    /// true if the current settings are valid
    pub fn PxJointLimitParameters_isValid(self_: *const physx_PxJointLimitParameters_Pod) -> bool;

    pub fn PxJointLimitParameters_isSoft(self_: *const physx_PxJointLimitParameters_Pod) -> bool;

    pub fn PxJointAngularLimitPair_new(lowerLimit: f32, upperLimit: f32, contactDist_deprecated: f32) -> physx_PxJointAngularLimitPair_Pod;

    pub fn PxJointAngularLimitPair_new_1(lowerLimit: f32, upperLimit: f32, spring: *const physx_PxSpring_Pod) -> physx_PxJointAngularLimitPair_Pod;

    /// Returns true if the limit is valid.
    ///
    /// true if the current settings are valid
    pub fn PxJointAngularLimitPair_isValid(self_: *const physx_PxJointAngularLimitPair_Pod) -> bool;

    pub fn PxJointAngularLimitPair_delete(self_: *mut physx_PxJointAngularLimitPair_Pod);

}

